### ISSUE TITLE
Allow usage of cmf bundle for non-PHPCR/cmf documents [RFC]

### ISSUE THREAD

**Issue by benglass** [symfony-cmf/SeoBundle#146]
With #145 being addressed I'd like to discuss ways to make this bundle usable for any kind of dynamic seo content management. It could be argued that this out of scope for this bundle but I dont think it would represent significant changes.

Right now there is a listener which looks for the cmf document in the request and updates the metadata based on it if it implements SeoContentAwareInterface.

A listener which goes through all of the attributes to find SeoContentAware items (that were converted by the paramconverter) is probably a bad idea due to the overhead it introduces to request.

I am thinking we could have an annotation that allows you to tag a parameter to a controller as an SeoContentObject. This would then hook into all the existing code for updating the seo presentation/sonata page. This would give you 2 options for having seo content populated by a dynamic object that implements SeoContentAware

1. Pull the seo presentation service out of the container and manually call updateSeoMetadata($object) in your controller action
2. Add an annotation that marks one of your arguments as a SeoContentObject or as SeoContentAware

```php
class FooController
{
    /*
     * @ParamConverter("foo")
     * @SeoContent("foo")
     */
    public function showAction(Foo $foo)
    {
        ...
    }
}
```

Thoughts?

I think having this bundle as a general solution to dynamic metadata management for symfony would be great as i dont think there is a bundle that addresses this (sonata does this for rendering meta tags but is missing the re-usable implementation for extracting seo metadata that this bundle provides). This kind of re-usability could make it a gateway of sorts to the other cmf packages.

**Comment by dbu** [comment]
ping @benglass

**Comment by benglass** [comment]
@dbu Ill take a look at putting together a PR with a twig function for setting seo metadata

**Comment by benglass** [comment]
@dbu is it the CMF coding standard/best practice to always create a helper service that is used for the twig extension? In this case I believe we will only have a single twig method of cmf_seo_update_seo_page but should I still put it inside a service injected into the twig extension a la media bundle?

https://github.com/symfony-cmf/MediaBundle/blob/master/Twig/Extension/CmfMediaExtension.php

**Comment by benglass** [comment]
Reviewing the media bundle and what kinds of tests exists for the twig extension it seems like one of the main reasons to do this is to make testing simpler (unit test the helper not the extension). The other benefit of making the helper re-usable doesnt apply just yet in this case because its going to be a single method proxying to SeoPresentation->updateSeoPage($content) and SeoPresentation is already available as a service. My inclination is to create the seoHelper for consistency and scalability.

**Comment by dbu** [comment]
sf has (or rather: had until 3.0) the php templating engine without twig.
main reason was to support that. @WouterJ should we keep doing this now sf
deprecated php templating?

**Comment by dbu** [comment]
@benglass is #263 the only thing missing for this issue? if so i would like to close this one as we discussed lots of things and its difficult to see what the action items would be. if there are other things, can we create separate new issues?

**Comment by benglass** [comment]
@dbu The only other approach I recall discussing to help people use this bundle for non PHPCR documents was perhaps an annotation to tag a controller action as the content to use for updateSeoPage. This could also be handled by an optional request listener which loops over $request->attributes looking for objects that implement SeoAwareInterface and pass them off to $seoPresentation->updateSeoPage

As it stands people will be able to use the bundle to set seo metadata from non PHPCR objects through 2 means:

* Use seo presentation service in your controller
* Use the twig helper

Unless we want to provide a more automatic way for this to happen for non PHPCR docs (like the way that the request listener checks cmfMainContent to see if its SeoAware) then I think we can close this once the twig extension is merged in.

**Comment by dbu** [comment]
#263

**Comment by dbu** [comment]
> perhaps an annotation to tag a controller action as the content to use
> for updateSeoPage. This could also be handled by an optional request
> listener which loops over $request->attributes looking for objects that
> implement SeoAwareInterface and pass them off to
> $seoPresentation->updateSeoPage

we currently have
http://symfony.com/doc/master/cmf/bundles/seo/introduction.html#the-contentlistener
so users can configure it for something different than cmfMainContent
and its document already.
i just updated that documentation a bit to make it more understandable
https://github.com/symfony-cmf/symfony-cmf-docs/pull/711

i think looping over all request attributes and doing instanceof checks
is really heavy. people should write their own, more clever listener in
that case imo.

oh, i actually already closed the issue when merging the PR you did. but
i think there is nothing left to do.

**Comment by benglass** [comment]
In order to use the content listener for non PHPCR docs you would have to either:

I agree looping over all request attributes is not a very good solution.

Here is another thought.

Right now since the content listener only supports checking 1 content key in the request, you would have register multiple copies of it with different content keys for every argument name in any of your actions that represents an SEO object unless you want to always call your action argument $seoContent which is not a desirable restriction. 

For example, what if you have a BlogPost object that implements SeoAware? You would have to register a separate content listener which looks at for a $blogPost key in your action. This would not scale very well.

If it supported checking multiple keys then we could allow the user to configure it to listen on 'cmfMainContent' as well as 'blogPost' and 'job' etc. This seems like it might be a nice enhancement actually to allow configuring multiple contentKeys for the content listener.

**Comment by dbu** [comment]
we will call the updateSeoPage with each thing we find. but usually we
should find only one, and even if we find more it should be no big deal
as the extractor has a supports() mechnism and will ignore stuff that is
not supported.

if you want to do a PR to allow the keys to be an array (and BC to make
a single key an array with 1 element) i think that is a good idea.

**Comment by benglass** [comment]
@dbu Created a separate issue with a summary, I'll work on a BC PR and this issue here and its epic discussion thread can be lost to the sands of time.
